added a BC: workspace document (#2197)
iterative/dvc.org: content/docs/user-guide/basic-concepts/workspace.md

name: Workspace
match: [workspace]
tooltip: >-
  The directory containing all your project files, e.g., the raw data, source
  code, ML models. Typically, it's also a Git repository. It contains your DVC
  project.
---

# Workspace

A data science project can consist of data obtained from many distinct sources.
These may be split into multiple files or directories or (as the project
structure needs) have different versions for different requirements, e.g., a
smaller / simplified version might be required in prototyping for faster

> feedback and shorter training times. A single workspace to manage all artifacts

Reviewer: > or have different versions for different requirements...

Let's not go into versioning here, I think. At least not by implying they're all in the workspace because in DVC the workspace only holds one version (the rest are cached and managed via Git, metafiles, etc.

Reviewer: > versioning needs and managing dependencies make it increasingly difficult

p.s. This is  better way to very subtly mention versioning (could even link to the corresponding Use Case doc).

of a project is desirable, although versioning needs and managing dependencies
make it increasingly difficult.

> DVC allows a single directory to contain all your project artifacts. The

Reviewer: > a single directory to contain all your project artifacts

Not exactly. File contents are in the cache with a special file structure (see https://dvc.org/doc/user-guide/project-structure/internal-files#structure-of-the-cache-directory)

> The workspace is the directory containing the _visible_ part of your project

That part is correct. And contradicts the previous part 🙂 (because "visible part" implies there's a hidden part which must be in other dirs).

Reviewer: > The workspace is the directory containing the visible part of your project

Let's open this p with that sentence.

workspace is the directory containing the _visible_ part of your

> <abbr>project</abbr>, e.g., the raw data, source code, model files. You can have

Reviewer: "e.g., the raw data, source code, model files **you're currently using**"

multiple versions of data, models, and other kinds of artifacts within the

> workspace and limit your focus to a subset of these. You can record your

Reviewer: > You can have multiple versions of data ... within the workspace

Again contradicting 😕

> progress in a commit and analyze your data and model history. DVC provides a

Reviewer: > record your progress in a commit and analyze...

Again probably too many details about versioning. Doesn't really fall within the 'workspace' concept, I think. This can be simpler.

Reviewer: > No need to rename your models for minor changes...
or save tens of different renamed files for training

Those are better mentions of versioning (clear benefits i.e. how much you'd have to suffer without DVC)

> save cleaned up data in different directories

That specific example isn't great because that's still pretty common even with DVC (e.g. in our own example-get-started repo we have a `prepared/` dir).

> _machine learning file system_ to manipulate your data and models using its

Reviewer: This "ML file system" term is pretty tricky. No need to force it (just skip it if you can't find a correct way to use it).

I can only think of something like "DVC turns your project into a sort of machine learning file system for..." but not sure.

commands. No need to rename your models for minor changes, save cleaned up data
in different directories or save tens of different renamed files for training
programs. DVC can keep track of all of these in a single directory called the

> workspace.

Reviewer: > DVC can keep track of all of these in a single directory called the
workspace.

Again contradicting and also, repetitive at this point.

Files and directories in the workspace can be added to DVC (`dvc add`), or they
can be downloaded from external sources (`dvc get`, `dvc import`,
`dvc import-url`). Changes to the data, notebooks, models, and any related
machine learning artifact can be tracked (`dvc commit`), and their content can
be synchronized (`dvc checkout`). Tracked data can be removed (`dvc remove`)
from the workspace.

DVC supports all typical operations of a versioned data file system through its

> commands. Behind the scene these operations use <abbr>metafiles</abbr> like the

Reviewer: > DVC supports all typical operations of a versioned data file system through its commands.

Maybe open the **previous** paragraph with that?

p.s. having this I think def. no need for the "ml file system" keyword. But keeping "machine learning" somewhere would be nice.

Reviewer: Let's try to incorporate the metafile mentions to the main (2nd) paragraph somehow. After simplifying it per my previous comments, there should be enough room in there. that way there's no need for this 4th p.

  models, etc. Typically, it's also a Git repository. It will contain your DVC
  project.
---

# Workspace

A data science project consists of data obtained from many different sources.
This data may be split into multiple files or directories or (as the project
structure needs) have different versions for different requirements. e.g. A
smaller / simplified version might be required in prototyping for faster
feedback and shorter training times. A single workspace to manage all artifacts
of a project is desirable, although versioning needs and managing dependencies
make it increasingly complex.

DVC allows a single directory to contain all your project artifacts. The
workspace is the directory containing _user visible_ part of your

> <abbr>project</abbr> e.g. raw datasets, source code, ML models, etc. Users work

Reviewer: - _user visible_ -> just _visible_ ?
- "e.g. raw data..." -> "raw data, ... models, etc. that you're currently using." perhaps

Author: To my taste _visible_ is cleaner. But someone can say _visible to whom?_ to this and we may need to explain it. 

_Raw data_ is better, yes. 

Thanks.

Reviewer: "Visible part of the project" seems perfectly OK to me.

> Raw data is better

You already had "raw data" 🙂 My suggestion is to add "that you're currently using."

Author: I thought you meant dataset ➡️  data 😆 

"that you're currently using" seems a bit redundant to me but let's see what happens.

Reviewer: One of the points of DVC projects is to track multiple versions of data,models,etc. so pointing out that the workspace is where you find the current versions seems pretty key to me. But open to suggestions on other ways to emphasize that.

Author: How about the changes in 44af0a9f ?

Reviewer: Hey sorry for the delay... Reading now.

Reviewer: > can have multiple versions of data, models, and other kinds of artifacts within the workspace

This doesn't sound right. The workspace only has one version of the data, etc. at a time.

Reviewer: And the paragraph became really long again!

in this directory using their data and model files that and manipulate the
contents through DVC commands.

Files and directories in the workspace can be added to DVC (`dvc add`) or they

> can be downloaded from external sources (`dvc get`, `dvc import`,

Reviewer: > `dvc get`, `dvc import`, `dvc import-url`

Let's just list "import, get, etc." I think.

Author: I repeat `dvc` part for links. Is there a way to link implicitly without `dvc`? Otherwise I'll link them manually.

Reviewer: Yes, def. keep `dvc`. I meant basically to remove the `import-url` link (kind of redundant with `import`).

Reviewer: But actually `add` can also download data to the workspace (see `--out` and `--to-remote` options). Also, `import*` commands download AND track data. You may want to rephrase this part accordingly 🙂

Author: Sometimes I think there is no command that hasn't got a duplicate, somehow :) I try to mention commands in passing, if we'd consider each and every option to commands, we'll need to duplicate the command reference here IMHO. 

We may just delete the commands if you would like.

Author: I can also list all possibilities for each functionality, like

In the workspace, you can
- Import files and directories (`dvc add --out`, `dvc import-url`...)

but I think this will turn the document into a list of commands and options.

Reviewer: No need to list every command usage of course, agreed!

My point was that these 3 commands mentioned actually overlap in a way that makes the current text slightly incorrect. In any case, the main use case of `add` is not to "add" but to "track", actually. Please check each cmd ref to try to find the right terms when needed 🙂 

"Download" is correct for `get/import` but `add` can also download (and they can all "transfer") so I'd avoid that term probably. And in fact I wouldn't even mention `get` here, since it doesn't require a DVC project/workspace. For `import` I'd try to use the cmd name as the relevant action (to "import") I guess...

Reviewer: > the main use case of add is not to "add" but to "track"
> For import I'd try to use the cmd name as the relevant action (to "import")

then again `import*` also track the downloaded data 😅. Maybe it should be a single sentence about tracking and put both `add`, `import`, `import-url` in the same parenthesis.

> `dvc import-url`). Changes to the files, directories, notebooks, models, and in

Reviewer: We already listed "files and dirs" so maybe just "data" here?

Author: This should be done in 42559e2.

Reviewer: Yep, thanks.

> general machine learning file system can be tracked (`dvc commit`) and versioned

Reviewer: I know "machine learning file system" was a desired SEO term (i.e. something people search for) but the way it's used here doesn't make sense IMO. What are ML FSs? (Idk)

I would skip that part here ("and in general...") and look for another way to use that keyword, if possible.

Author: I moved the phrase to the previous paragraph and modified this in 44af0a9 and 42559e2

Reviewer: Better! 👍

> in Git (`dvc checkout`). They can be removed (`dvc remove`) from the workspace.

Reviewer: - It sounds like we're saying that `dvc checkout` versions data with Git. Can you clarify a bit?
- "They can be removed" -> "Tracked data can be..."

Author: Modified in 6cc0f6c

Reviewer: > data, notebooks, models, and any related machine learning artifact

That list got a little long? Prob no need for "machine learning" here since the term is found elsewhere in the doc...

> their content can be synchronized (`dvc checkout`)

Good one! But needs some clarifying (sync what)?

Also, people usually need checkout before commit (which is included in add/repro). So maybe something like:

"When switching Git revisions in your |repo|, you can use `dvc checkout` to sync |metafiles| with the other workspace files and directories. If you manually modify tracked data, use `dvc commit` to record the changes with DVC. And if needed, tracked data can be removed..."

Author: > "When switching between |repository| versions, use dvc checkout to sync DVC-tracked data with Git-tracked |metafiles|. If you manually modify the workspace status, use dvc commit to record the changes. And if needed, tracked data can be removed..."

This sentence looks more suitable to use cases or the user guide. Here I would like to mention just the capabilities. Conceptually `dvc checkout` is similar to `git checkout` and `dvc commit` is akin to. `git commit`. This can be turned into a bullet list maybe,

- You can add your files and directories ...
- You can download external sources
- You can synchronize your workspace with the cache
- You can do this
- You can do that

I feel this paragraph is too dense.

Author: But reading again, this looked rather like a political ad than a technical documentation. Maybe "you can"s are spurious. Just 

- Add your files (`dvc add`)
- Download external sources (`dvc get`...)
- ...

is better.

<abbr>Pipelines</abbr> and <abbr>dependencies</abbr> between them can be
defined. Data and model files can be moved to the cloud and retrieved when

> necessary (`dvc push`, `dvc pull`). DVC supports all typical operations of files

Reviewer: I don't see a direct relationship between pipelines, dependencies, remotes (push/pull) and workspace. No need to force all DVC features into this paragraph 🙂

Author: 👍🏼  I'm removing those two sentences in 6cc0f6c

> and directories of a file system through its commands.

Reviewer: > DVC supports all typical operations...

What do we mean by this?

Author: We mean the user can live in the workspace and use dvc to do what they normally do with files. Create, copy, rename etc.

Reviewer: I see. Yes it may be a relevant note. This sentence seems to belong more to the previous paragraph though? Something like:

"There's usually no need to modify tracked data manually as DVC commands provide ways to do so, but if you do, use `dvc commit` to register the changes..."

(see my previous comment, these would have to be merged somehow).

Behind the scene these operations of a <abbr>DVC project</abbr> uses
<abbr>metafiles</abbr> like the `.dvc/` directory, `dvc.yaml` or files with

> `*.dvc` extension to track the content and dependencies.

Reviewer: these -> the?

But I'm still not sure I'm getting how this p related to the concept of workspace. May just need some rephrasing because mentioning dvc.yaml and *.dvc files could def. be relevant. Or fit those files into the intro p (probably best).

Author: _these_ refer to _typical operations_ from the previous paragraph.

Author: I think introducing `.dvc` files in the intro is a bit distracting. 

Could you check e85a0d76 for changes to this?

Reviewer: I think we can definitely incorporate .dvc files and dvc.yaml mentions earlier, closer to the main concept definition. Metafiles are one of the most important contents of the workspace, along with the corresponding data. That, the corresponding data, and any git-tracked assets (mainly code). .dvc/ is not considered part of the workspace. Workspace is analogous to working tree in Git.

  models, etc. Typically, it's also a Git repository. It will contain your DVC
  project.
---

<!-- keywords: data science project architecture, machine learning project architecture, machine learning workflow, data science workflow, machine learning file system, data science file system, data science project structure, machine learning project structure, notebook version control -->

# Workspace

A data science project consists of data obtained from many different sources.
This data may be split into multiple files or directories or (as the project
structure needs) have different versions for different requirements. e.g. A
smaller / simplified version might be required in prototyping for faster
feedback and shorter training times. A single workspace to manage all artifacts
of a project is desirable, although versioning needs and managing dependencies
make it increasingly complex.

> DVC allows a single directory to contain all your project artifacts. In the

Reviewer: What's the main point of this part of the story/ motivation? I'm not sure what we want to say here. I'm starting to think "workspace" is kind of a plain concept which may not need much introduction at all.

I know in part you're trying to fit a bunch of SEO words but you were right that those are secondary. Let's get the story right first I think, and then sprinkle special terms if possible.

Reviewer: I think that mentioning directory structures could definitely make sense. Just not sure whether that comes before or after defining what the workspace is. Probably after in this case?

Author: Directory structure of the workspace or `.dvc/` I think a separate _DVC project_ (or repository) BC document is more suitable for this.

Reviewer: No, I mean "data may be split into multiple files or directories" etc. — how to structure files in the workspace dir.

Reviewer: p.s. there's an entire guide for the [spec of .dvc/](https://dvc.org/doc/user-guide/project-structure/internal-files) already, no need to add that to the project concept page 🙂

Author: I can tell these in a "there are such and such options to organize your data and models" way I think. It looks too context/project dependent to have a clear recommendation. Bottom line: User can organize their data in any way they like 😄 

Is there a best practice document I can take a look at for this? I can probably cook some set of examples, is this what you would like?

Reviewer: > too context/project dependent to have a clear recommendation... organize their data in any way they like

OK. So why did you include that info in the first place? "data may be split into multiple files or directories..."

> best practice document I can take a look at for this? I can probably cook some set of examples

Let's please focus on this issue (write the workspace concept page) for now. There's a separate epic issue for Best Practices somewhere but again, let's not go there now 🙂

Author: A data science project can have different kinds of data, from different sources, in different directories distributed in different machines. Workspace is where we collect them together. I mean this. 

Could you please check the changes in 0b099f6. Thank you 😄

Reviewer: OK. I see you rewrote the 2nd paragraph. This comment refers to the first paragraph, the motivation/intro.

> A data science project can have different kinds of data, from different sources, in different directories distributed in different machines.

OK that's clear and concise. Why not use something direct and short like that for the motivation part? May not even need a whole paragraph 🙂

As it is now you need to read a whole paragraph and then another opening sentence (in the next p) to get to the "what is workspace" concept definition.

Again, just trying to filter out information that may be distracting/confusing/repetitive, even all the details everything you're writing is important in other contexts. If editing current text is becoming difficult I recommend you put this on a side and start over, then compare both versions if needed. Thanks

> DVC allows a single directory to contain all your project artifacts. In the
> documentation the workspace is the _user visible_ part of the directory that

Reviewer: ```suggestion
DVC allows a single directory to contain all your project artifacts.
the workspace is the _user visible_ part of the directory that
```

> DVC allows a single directory to contain all your project artifacts. In the
> documentation the workspace is the _user visible_ part of the directory that
> contains all your <abbr>project</abbr> files e.g. raw datasets, source code, ML

Reviewer: ```suggestion
DVC allows a single directory to contain all your project artifacts.
The workspace is the directory containing the _user visible_ part of
your <abbr>project</abbr> e.g. raw datasets, source code, ML
```

Author: Fixed in 50964b1 

Thank you 👍🏼

Reviewer: Np. These copy edits can be easily committed from GH btw but as you prefer.

models, etc. Users work in this directory using their data and model files that
and manipulate the contents through DVC commands.

Files and directories in the workspace can be added to DVC (`dvc add`) or they
can be downloaded from external sources (`dvc get`, `dvc import`,
`dvc import-url`). Changes to the files, directories, notebooks, models and in
general machine learning file system can be tracked (`dvc commit`) and versioned
in Git (`dvc checkout`). <abbr>Pipelines</abbr> and <abbr>dependencies</abbr>
between them can be defined. Data and model files can be moved to the cloud
(`dvc remote`) and retrieved one by one when necessary (`dvc push`, `dvc pull`).
They can be removed (`dvc remove`) from the workspace. DVC supports all typical
operations of files and directories of a file system through its commands.

> 
> Behind the scene these operations of a <abbr>DVC project</abbr> uses
> <abbr>metafiles</abbr> to track the content and dependencies.

Reviewer: Good idea to mention these but maybe mention .dvc and dvc.yaml files specifically instead, as examples of what the workspace can contain (in the main paragraph).

```suggestion
```

Author: Mentioning files by name makes me nervous 😅 

I added the names to the last paragraph in 46fe4bba but if you think it's better to remove the paragraph completely, I can just do that as well. Thanks.

Reviewer: Will make a seprate comment. Resolving this one.

  models, etc. Typically, it's also a Git repository. It will contain your DVC
  project.
---

> <!-- keywords: data science project architecture, machine learning project architecture, machine learning workflow, data science workflow, machine learning file system, data science file system, data science project structure, machine learning project structure, notebook version control -->

Reviewer: Just a note to remove this comment later before merging (but for now it's useful to have the list during review 👍)

Reviewer: Great job incorporating all of the terms BTW. We may have to drop a few but it's best to start having too many. Only "notebook version control" is missing but TBH I don't see how having that in workspace helps much — seems somewhat unrelated.

Author: 😄 Actually the first two paragraphs are added yesterday. [Last week's workspace document](https://github.com/iterative/dvc.org/blob/ug-basic-concepts-feb-21/content/docs/user-guide/concepts/workspace.md) does not contain them. I can simply remove them. I also think they are not related.

# Workspace

A data science project consists of data obtained from many different sources.

> Most of the time it needs to convert the the format of this data into a form

Reviewer 2: the the

Author: fixed. thanks.

that is required by the training models and supplying to data science / machine

> learning workflows. Sometimes it needs to be split into multiple files or

Reviewer: OK to have some context but it should probably be a short intro paragraph only? Maybe just say something about project structures.

Reviewer: BTW try to use other tooltips when possible e.g. `<abbr>project</abbr>` in this case.

Author: I think in BC documents, we need to make separation of concerns. If our aim is a quick introduction to DVC concepts, we don't need such *story material* to begin with. This was solely for SEO purposes. 

IMHO I can write blog posts / tutorials and tell stories with all kinds of SEO phrases there instead of trying to use these in docs. As you can see it's possible but I don't think it increases the quality of docs. 

I can paraphrase into a single paragraph or remove them completely. I'm fine with all.

Author: I didn't want to put `<abbr>project</abbr>` links to these as they _may_ (or may not) divert the reader. I was intending to define workspace first and then tell them about *DVC project.* 

![Screen Shot 2021-02-16 at 19 46 30](https://user-images.githubusercontent.com/476310/108094773-3b742100-7090-11eb-966e-bf67a81a02e9.png)

I'm putting <abbr> link to the last of these highlighted _project_ words.

Reviewer: > we need to make separation of concerns. If our aim is a quick introduction to DVC concepts, we don't need such story material

We do already have tooltips for shot definitions. But you can't search for "pipeline dvc" on Google and find a tooltip (i.e. we want to have landing pages for our docs).

Another goal of these is to extract explanations out of the cmd refs (e.g. in that search example you find https://dvc.org/doc/command-reference/dag which shouldn't be where pipelines are explained)

In my mind most concept pages are 2-3 paragraph long in general, I think (we'll see as they come). There could also be some cool diagrams potentially (feel free to contribute sketches, we can fix them up later). They can still have a story, maybe just a very straightforward one? (motivation -> explanation -> uses/implications).
WDYT @shcheklein ?

> I'm putting link to the last of these highlighted project words.

I usually "tooltipy" just one instance of the term per page or big section, but I try to make it the first or 2nd one (as long as it doesn't distract e.g. no other links are nearby).

Reviewer: p.p.s `straightforward one? (motivation -> explanation -> uses/implications` - I see that's more or less the structure here already 👍

directories or (as the project structure needs) have different versions for
different requirements. e.g. A smaller / simplified version might be required in
protoyping for faster feedback and shorter training times.

These activities all involve to change the machine learning / data science
project structure in unforeseeable ways and tracking them manually becomes
tedious. A single workspace to manage all artifacts of a project is desirable,
although versioning needs and managing dependencies make it increasingly
complex.

DVC allows a single directory to contain all your project artifacts. In the
documentation the workspace is the _user visible_ part of the directory that
contains all your project files e.g. raw datasets, source code, ML models, etc.

> Users work in this directory using their data and model files that and

Reviewer: > the **workspace** is the _user visible_ part of the **DVC project**

I would try to get to this rather quickly (after a short intro p if needed). This is the main content.

manipulate the contents through DVC commands.

Files and directories in the workspace can be added to DVC (`dvc add`) or they
can be downloaded from external sources (`dvc get`, `dvc import`,
`dvc import-url`). Changes to the files, directories, notebooks, models and in
general machine learning file system can be tracked (`dvc commit`).
<abbr>Pipelines</abbr> and <abbr>dependencies</abbr> between them can be defined
(`dvc run`). Data and model files can be sent to or retrieved from
<abbr>remotes</abbr> (`dvc push`, `dvc pull`) and can be removed (`dvc remove`,

> `dvc gc`) from the workspace. DVC supports all typical operations of files and

Reviewer: Good idea for the secondary part of concept pages: mention/link to the most relevant commands and features. In this case I agree about `add/import*/remove/move`, and `commit`, but importantly, we're currently missing `checkout`. Maybe `push/pull` (not sure, maybe `checkout` is enough as it relates workspace and cache). Probably not `gc` or `run`.

Reviewer: > cleanup operations are also a major hurdle for the common ML engineer

Sure but `gc` doesn't operate on the workspace, but rather on the cache (or remote). The workspace is always kept clean automatically (you only see 1 version at a time).

Def OK on `remove`, in fact it should probably be closer to the top (:

Similarly `push/pull` don't relate that much to the workspace, they're mostly operations between cache and remote. OK `pull` a little bit but it could be a parenthesis next to `checkout` probably.

Reviewer: I'm also not seeing much need for the `remote` link BTW.

Author: This is addressed in bdbcc97
